src/grisp_connect_client.erl

```diff
@@ -147,19 +106,57 @@ handle_event(state_timeout, wait, connecting, Data) ->
             ?LOG_INFO(#{event => waiting_ws_connection}),
             {keep_state_and_data, [{state_timeout, ?STD_TIMEOUT, wait}]}
     end;
-handle_event(cast, disconnected, connecting, _Data) ->
+connecting(cast, disconnected, _Data) ->
     repeat_state_and_data;
+?HANDLE_COMMON.
 
-% CONNECTED
-handle_event(enter, _OldState, connected, _Data) ->
+connected(enter, _OldState, _Data) ->
     grisp_connect_log_server:start(),
     keep_state_and_data;
-handle_event(cast, disconnected, connected, Data) ->
+connected({call, From}, is_connected, _) ->
+    {keep_state_and_data, [{reply, From, true}]};
+connected(cast, disconnected, Data) ->
     ?LOG_WARNING(#{event => disconnected}),
     grisp_connect_log_server:stop(),
     {next_state, waiting_ip, Data};
+connected(cast, {handle_message, Payload}, #data{requests = Requests} = Data) ->
+    Replies = grisp_connect_api:handle_msg(Payload),
+    % A reduce operation is needed to support jsonrpc batch comunications
+    case Replies of
+        [] ->
+            keep_state_and_data;
+        [{request, Response}] -> % Response for a GRiSP.io request
+            grisp_connect_ws:send(Response),
+            keep_state_and_data;
+        [{response, ID, Response}] -> % GRiSP.io response
+            {OtherRequests, Actions} = dispatch_response(ID, Response, Requests),
+            {keep_state, Data#data{requests = OtherRequests}, Actions}
+    end;
+connected({call, From}, {request, Method, Type, Params},
+          #data{requests = Requests} = Data) ->
+    {ID, Payload} = grisp_connect_api:request(Method, Type, Params),
+    grisp_connect_ws:send(Payload),
+    NewRequests = Requests#{ID => From},
+    {keep_state,
+     Data#data{requests = NewRequests},
+     [{{timeout, ID}, request_timeout(), request}]};
+?HANDLE_COMMON.
 
-handle_event(E, OldS, NewS, Data) ->
+% Common event handling appended as last match case to each state_function
+handle_common(cast, connect, _State, _Data)->
+    keep_state_and_data;
+handle_common({call, From}, is_connected, _, _) ->
+        {keep_state_and_data, [{reply, From, false}]};
+handle_common({call, From}, {request, _, _, _}, _State, _Data) ->
+    {keep_state_and_data, [{reply, From, {error, disconnected}}]};
+handle_common({timeout, ID}, request, _, #data{requests = Requests} = Data) ->
+    Caller = maps:get(ID, Requests),
+    {keep_state,
+     Data#data{requests = maps:remove(ID, Requests)},
+     [{reply, Caller, {error, timeout}}]};
+handle_common({call, From}, _, State, Data) ->
+    {keep_state, Data, [{reply, From, {bad_client_state, State}}]};
+handle_common(E, OldS, NewS, Data) ->
```

Comment:
As a refactoring this is correct. But in general I would only tolerate unexpected info events (and log them) and crash on all other events, since those would mean there is a bug in the implementation.

Reply:
Mmm good point, indeed I put that there mainly to nicely catch edge cases during development. Should I remove it and let it crash?

Reply:
You can make the crashes more meaningful with something like:
```
handle_common(cast, Cast, _, _) ->
    error({unexpected_cast, Cast});
handle_common({call, _}, Call, _, _) ->
    error({unexpected_call, Call});
handle_common(info, Info, State, Data) ->
    ?LOG_ERROR(#{event => unexpected_info,
                 info => Info,
                 state => State,
                 data => Data}),
    keep_state_and_data.
```

One could discuss whether to crash on unexpected_info as well. I just think it's ok to allow processes to send random messages to other processes. There are still other events like unhandled timeouts or internal events that will crash with a case clause here, but since those bugs would be module internal and not only app internal, I think a common crash report is fine for that.